### PR TITLE
feat: delete tags

### LINKED ISSUE #5146: Delete tags
As part of the action menu on tagging we should also have:

- [ ] Delete tag: This would need a confirmation modal saying something like: "The tag will be deleted from any messages."

### PR DESCRIPTION
**Pull request by hamza221**
closes  #5146

image

- [x] fix UI : delete modal 
- [x] polish Pr
- [x] refresh tag list after deletion 
- [x] fix failing test

### CODE REVIEW

File: src/components/DeleteTagModal.vue [hunk 1]
```diff
@@ -68,11 +71,14 @@ export default {
 					tag: this.tag,
 					accountId: this.accountId,
 				})
+				this.$store.commit('deleteTag', {
```
Comment (ChristophWurst) [inline src/components/DeleteTagModal.vue:74]: move the mutation call into the action

File: lib/Contracts/IMailManager.php [hunk 2]
```diff
@@ -317,6 +317,17 @@ public function createTag(string $displayName, string $color, string $userId): T
 	 */
 	public function updateTag(int $id, string $displayName, string $color, string $userId): Tag;
 
+	/**
+	 * Update a mail tag
+	 *
+	 * @param string $userId
```
Comment (kesselb) [inline lib/Contracts/IMailManager.php:323]:
```suggestion
```

File: lib/Contracts/IMailManager.php [hunk 3]
```diff
@@ -317,6 +317,17 @@ public function createTag(string $displayName, string $color, string $userId): T
...
+	 * @param int $id
```
Comment (kesselb) [inline lib/Contracts/IMailManager.php:324]:
```suggestion
```

File: lib/Contracts/IMailManager.php [hunk 4]
```diff
@@ -317,6 +317,17 @@ public function createTag(string $displayName, string $color, string $userId): T
...
+	 *
+	 * @return int
```
Comment (kesselb) [inline lib/Contracts/IMailManager.php:326]:
```suggestion
```

File: lib/Controller/TagsController.php [hunk 5]
```diff
@@ -27,21 +27,28 @@
...
 	private string $currentUserId;
 	private IMailManager $mailManager;
 
+	private AccountService $accountService;
+
+
 	public function __construct(IRequest $request,
 		string $UserId,
-		IMailManager $mailManager
+		IMailManager $mailManager, AccountService $accountService,
```
Comment (kesselb) [inline lib/Controller/TagsController.php:46]:
```suggestion
		IMailManager $mailManager, 
		AccountService $accountService,
```

File: lib/Db/MessageTags.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+use OCP\AppFramework\Db\Entity;
+use ReturnTypeWillChange;
+
+/**
+ * @method string getImapMessageId()
+ * @method void setImapMessageId(string $imapMessageId)
+ * @method int getTagId()
+ * @method void setTagId(int $tagId)
+ */
+class MessageTags extends Entity implements JsonSerializable {
```
Comment (kesselb) [inline lib/Db/MessageTags.php:38]: Please add a constructor with the addType definition for tagId 

```
	public function __construct() {
		$this->addType('tagId', 'integer');
	}
```

File: lib/Db/MessageTagsMapper.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+use OCP\IL10N;
+
+/**
+ * @template-extends QBMapper<MessageTags>
+ */
+class MessageTagsMapper extends QBMapper {
+	/** @var IL10N */
+	private $l10n;
```
Comment (kesselb) [inline lib/Db/MessageTagsMapper.php:38]:
```suggestion
```

File: lib/Db/MessageTagsMapper.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+	public function __construct(IDBConnection $db,
+		IL10N $l10n) {
+		parent::__construct($db, 'mail_message_tags');
+		$this->l10n = $l10n;
```
Comment (kesselb) [inline lib/Db/MessageTagsMapper.php:43]:
```suggestion
```

File: lib/Db/MessageTagsMapper.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * @template-extends QBMapper<MessageTags>
+ */
+class MessageTagsMapper extends QBMapper {
+	/** @var IL10N */
+	private $l10n;
+
+	public function __construct(IDBConnection $db,
+		IL10N $l10n) {
```
Comment (kesselb) [inline lib/Db/MessageTagsMapper.php:41]:
```suggestion
		) {
```

File: lib/Contracts/IMailManager.php [hunk 10]
```diff
@@ -317,6 +317,18 @@ public function createTag(string $displayName, string $color, string $userId): T
...
 	public function updateTag(int $id, string $displayName, string $color, string $userId): Tag;
 
+	/**
+	 * Delete a mail tag
+	 *
+	 * @param int $id
+	 * @param string $userId
+	 * @param Account $account
+	 *
+	 * @return Tag
```
Comment (kesselb) [inline lib/Contracts/IMailManager.php:327]:
```suggestion
```

The phpdoc for param & return is only duplicates the method signature but does not provide additional information and therefore should go. 

Please keep the method description and throws.

File: lib/Controller/TagsController.php [hunk 11]
```diff
@@ -80,6 +88,24 @@ public function update(int $id, string $displayName, string $color): JSONRespons
 		$tag = $this->mailManager->updateTag($id, $displayName, $color, $this->currentUserId);
 		return new JSONResponse($tag);
 	}
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @param int $id
+	 *
+	 * @return JSONResponse
```
Comment (kesselb) [inline lib/Controller/TagsController.php:96]:
```suggestion
```

File: src/components/TagItem.vue [hunk 12]
```diff
@@ -79,6 +90,9 @@ export default {
 		}
 	},
 	methods: {
+		deleteTag() {
+			this.$emit('delete-tag', this.tag)
```
Comment (kesselb) [inline src/components/TagItem.vue:94]:
```suggestion
			this.$emit('deleteTag', this.tag)
```

Recommendation from https://v3.vuejs.org/guide/component-custom-events.html#event-names

File: lib/Controller/TagsController.php [hunk 13]
```diff
@@ -80,6 +88,21 @@ public function update(int $id, string $displayName, string $color): JSONRespons
...
 	}
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @throws ClientException
+	 */
+	#[TrapError]
+	public function delete(int $id, int $accountId): JSONResponse {
+		try {
+			$account = $this->accountService->find($this->currentUserId, $accountId);
```
Comment (kesselb) [inline lib/Controller/TagsController.php:99]: The message tags are shared for a user and therefore delete tag should remove it for every account.

File: lib/Service/MailManager.php [hunk 14]
```diff
@@ -493,7 +499,30 @@ public function flagMessage(Account $account, string $mailbox, int $uid, string
 			)
 		);
 	}
+	public function tagMessageWithClient(Horde_Imap_Client_Socket $client, Account $account, Mailbox $mailbox, Message $message, Tag $tag, bool $value):void {
 
+		if ($this->isPermflagsEnabled($client, $account, $mailbox->getName()) === true) {
+			try {
+				if ($value) {
+					// imap keywords and flags work the same way
+					$this->imapMessageMapper->addFlag($client, $mailbox, [$message->getUid()], $tag->getImapLabel());
```
Comment (kesselb) [inline lib/Service/MailManager.php:508]: addFlag / removeFlag accepts a list of uids. 

Do you think we can group the messages by mailbox in MailManager.deleteTag and pass a list of Messages to tagMessageWithClient?

The advantage is that we reduce the number of imap commands.

**Review by st3iny — COMMENTED** [review]
Tested and works.

If a tag is deleted that is assigned to the currently opened email, its tag list (below the header) won't be updated immediately. Changing to another email and back again updates the list accordingly.

I don't consider this a hard blocker.


**PS:** Eslint and psalm are not happy yet.

```
/home/runner/work/mail/mail/src/components/TagItem.vue
Error:   94:15  error  Custom event name 'deleteTag' must be kebab-case  vue/custom-event-name-casing
```

File: lib/Service/MailManager.php [hunk 15]
```diff
@@ -493,7 +499,37 @@ public function flagMessage(Account $account, string $mailbox, int $uid, string
...
+		if ($value) {
+			foreach ($messages as $message) {
+				$this->tagMapper->tagMessage($tag, $message->getMessageId(), $account->getUserId());
+			}
+		} else {
+			foreach ($messages as $message) {
+				$this->tagMapper->untagMessage($tag, $message->getMessageId());
+			}
+		}
+		}
```
Comment (ChristophWurst) [inline lib/Service/MailManager.php:531]: indentation is wonky

File: lib/Service/MailManager.php [hunk 16]
```diff
@@ -819,6 +855,50 @@ public function updateTag(int $id, string $displayName, string $color, string $u
...
+			} else {
+				$groupedMessages[$mailboxId] = [$message];
+			}
+		}
+
+		foreach ($groupedMessages as $mailboxId => $messages) {
+			$mailbox = $this->getMailbox($userId, $mailboxId);
+			$this->tagMessageWithClient($client, $account, $mailbox, $messages, $tag, false);
+		}
+		$client->logout();
```
Comment (ChristophWurst) [inline lib/Service/MailManager.php:900]: use the client with a try-finally to ensure the connection is closed on error too

**Review by st3iny — APPROVED** [review]
Tested and works.

**Pull request merged**